File: eel-core/src/main/kotlin/io/eels/component/jdbc/JdbcInserter.kt
package io.eels.component.jdbc

import io.eels.Row
import io.eels.component.ResultSetIterator
import io.eels.schema.Schema
import io.eels.util.Logging
import io.eels.util.zipWithIndex
import java.sql.DriverManager

class JdbcInserter(val url: String,
                   val table: String,
                   val schema: Schema,
                   val autoCommit: Boolean,
                   val dialect: JdbcDialect) : Logging {

  val conn = DriverManager.getConnection(url)

  init {
    logger.debug("Connecting to jdbc $url...")
    conn.autoCommit = autoCommit
    logger.debug("Connected to $url")
  }

  fun insertBatch(batch: List<Row>): Unit {
    val stmt = conn.prepareStatement(dialect.insertQuery(schema, table))
    try {
      batch.forEach { row ->
        row.values.zipWithIndex().forEach {
          stmt.setObject(it.second + 1, it.first)
        }
        stmt.addBatch()
      }
      val result = stmt.executeBatch()
      logger.debug("Batch completed; ${result.size} rows inserted")
      if (!autoCommit) conn.commit()
    } catch (e: Exception) {
      logger.error("Batch failure", e)
      if (!autoCommit)
        conn.rollback()
      throw e
    } finally {
      stmt.close()
    }
  }

  fun ensureTableCreated(): Unit {
    logger.debug("Ensuring table [$table] is created")

    fun tableExists(): Boolean {
      logger.debug("Fetching list of tables to detect if $table exists")
      val tables = ResultSetIterator(conn.metaData.getTables(null, null, null, arrayOf("TABLE"))).asSequence().toList()
      val tableNames = tables.map { it.get(3).toLowerCase() }
      val exists = tableNames.contains(table.toLowerCase())
      logger.debug("${tables.size} tables found; $table exists == $exists")
      return exists
    }

    if (!tableExists()) {
      val sql = dialect.create(schema, table)
      logger.info("Creating table $table [$sql]")
      val stmt = conn.createStatement()
      try {
        stmt.executeUpdate(sql)
        if (!autoCommit) conn.commit()
      } catch(e: Exception) {
        logger.error("Batch failure", e)
        if (!autoCommit)
          conn.rollback()
        throw e
      } finally {
        stmt.close()
      }
    }
  }
}

File: eel-core/src/main/scala/io/eels/component/jdbc/JdbcInserter.scala
package io.eels.component.jdbc

import java.sql.Connection

import com.sksamuel.exts.Logging
import com.sksamuel.exts.io.Using
import com.sksamuel.exts.jdbc.ResultSetIterator
import io.eels.Row
import io.eels.component.jdbc.dialect.JdbcDialect
import io.eels.schema.StructType

class JdbcInserter(val connFn: () => Connection,
                   val table: String,
                   val schema: StructType,
                   val autoCommit: Boolean,
                   val batchesPerCommit: Int,
                   val dialect: JdbcDialect) extends Logging with Using {

  logger.debug("Connecting to JDBC to insert.. ..")
  private val conn = connFn()
  conn.setAutoCommit(autoCommit)
  logger.debug(s"Connected successfully; autoCommit=$autoCommit")

  private var batches = 0

  def insertBatch(batch: Seq[Row]): Unit = {
    val stmt = conn.prepareStatement(dialect.insertQuery(schema, table))
    try {
      batch.foreach { row =>
        row.values.zipWithIndex.foreach { case (value, k) =>
          dialect.setObject(k, value, row.schema.field(k), stmt, conn)
        }
        stmt.addBatch()
      }
      batches = batches + 1
      stmt.executeBatch()
      if (!autoCommit) conn.commit()
      else if (batches == batchesPerCommit) {
        batches = 0
        conn.commit()
      }
    } catch {
      case t: Throwable =>
        logger.error("Batch failure", t)
        if (!autoCommit)
          conn.rollback()
        throw t
    } finally {
      stmt.close()
    }
  }

  def dropTable(): Unit = using(conn.createStatement)(_.execute(s"DROP TABLE IF EXISTS $table"))

  def tableExists(): Boolean = {
    logger.debug(s"Fetching list of tables to detect if $table exists")
    val tables = ResultSetIterator.strings(conn.getMetaData.getTables(null, null, null, Array("TABLE"))).toList
    val tableNames = tables.map(x => x(3).toLowerCase)
    val exists = tableNames.contains(table.toLowerCase())
    logger.debug(s"${tables.size} tables found; $table exists == $exists")
    exists
  }

  def ensureTableCreated(): Unit = {
    logger.info(s"Ensuring table [$table] is created")

    if (!tableExists()) {
      val sql = dialect.create(schema, table)
      logger.info(s"Creating table $table [$sql]")
      val stmt = conn.createStatement()
      try {
        stmt.executeUpdate(sql)
        if (!autoCommit) conn.commit()
      } catch {
        case t: Throwable =>
          logger.error("Batch failure", t)
          if (!autoCommit)
            conn.rollback()
          throw t
      } finally {
        stmt.close()
      }
    }
  }
}